spring-integration/spring-boot-autoconfigure/src/main/java/ai/timefold/solver/spring/boot/autoconfigure/IncludeAbstractClassesEntityScanner.java

```diff
@@ -0,0 +1,115 @@
+package ai.timefold.solver.spring.boot.autoconfigure;
+
+import static java.util.Collections.emptyList;
+
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import ai.timefold.solver.core.api.domain.entity.PlanningEntity;
+import ai.timefold.solver.core.api.domain.solution.PlanningSolution;
+
+import org.springframework.beans.factory.annotation.AnnotatedBeanDefinition;
+import org.springframework.beans.factory.config.BeanDefinition;
+import org.springframework.boot.autoconfigure.AutoConfigurationPackages;
+import org.springframework.boot.autoconfigure.domain.EntityScanPackages;
+import org.springframework.boot.autoconfigure.domain.EntityScanner;
+import org.springframework.context.ApplicationContext;
+import org.springframework.context.annotation.ClassPathScanningCandidateComponentProvider;
+import org.springframework.core.type.AnnotationMetadata;
+import org.springframework.core.type.filter.AssignableTypeFilter;
+import org.springframework.util.ClassUtils;
+
+public class IncludeAbstractClassesEntityScanner extends EntityScanner {
+
+    private final ApplicationContext context;
+
+    public IncludeAbstractClassesEntityScanner(ApplicationContext context) {
+        super(context);
+        this.context = context;
+    }
+
+    public <T> Class<? extends T> findFirstImplementingClass(Class<T> targetClass) {
+        return Optional.ofNullable(findImplementingClassList(targetClass)).filter(classes -> !classes.isEmpty())
+                .map(classes -> classes.get(0)).orElse(null);
```

Comment:
Let's avoid `Optional` if the only purpose of it is to avoid a local `null` check.

I'd argue the following code reads much better:

    var implementingClassList = findImplementingClassList(targetClass);
    if (implementingClassList == null || implementingClassList.isEmpty()) {
        return null;
    } else {
        return implementingClassList.get(0);
    }
    
`Optional` is useful to communicate nullness information with outside code. But if we can keep the `null` local inside of a method, using `Optional` is just a more difficult way to write a `null` check.

Also, why are we bothering with `null` at all? `findImplementingClassList(...)` doesn't seem to ever return `null`.

Reply:
Okay, I'll remember that for next time. Also, you're right about not bothering with the value since it will never be null.